Handler saga registration (#8214)
Particular/docs.particular.net: nservicebus/handlers-and-sagas-registration_registration-content_core_[10,).partial.md

## Source-generated registration (recommended)

Starting in NServiceBus version 10.2, a source generation package can scan assemblies at build time and produce a strongly typed registration API. This keeps NServiceBus registrations componentized without requiring runtime assembly scanning.

The generated API is composable: each scanned assembly produces extension methods that attach to a registry. For example, after referencing the source generator package, the registration code looks like this:

```csharp
endpointConfiguration.Handlers.MyAssembly.MyNamespace.AddAll();
```

> Source-generated registration is trimming and AOT-friendly because it references handler and saga types directly from the composition root (the host). It also eliminates runtime surprises: any type that is discoverable at build time is discoverable at runtime with no additional scanning.

Reviewer: ```suggestion
Source-generated registration is compatible with assembly trimming and ahead-of-time (AOT) compilation because it references handler and saga types directly from the composition root, which is the host. It also eliminates runtime surprises: any type that is discoverable at build time is discoverable at runtime with no additional scanning.
```

> Source-generated registration is trimming-friendly and AOT-friendly because it references handler and saga types directly from the composition root (the host). It also eliminates runtime surprises: any type that is discoverable at build time is discoverable at runtime with no additional scanning.

Reviewer 2: ```suggestion
Source-generated registration is trimming and AOT-friendly because it references handler and saga types directly from the composition root (the host). It also eliminates runtime surprises: any type that is discoverable at build time is discoverable at runtime with no additional scanning.
```

### Advanced source generation configuration

The `[HandlerRegistryExtensions]` attribute can be applied to a `partial static class` to customize the source-generated handler registration API:

#### `EntryPointName`

Overrides the default assembly-based entry point name. The value must be a valid C# identifier.

```csharp
[HandlerRegistryExtensions(EntryPointName = "Shipping")]
public static partial class RegistrationExtensions;
```

> When the attribute above is applied, `registry.Handlers.Core...` becomes `registry.Handlers.Shipping...`.

Reviewer: "Core" is going to be confusing there. Where did "Core" come from? Would this be better?
```suggestion
When the attribute above is applied, `registry.Handlers.MyAssemblyName...` becomes `registry.Handlers.Shipping...`.
```

#### `RegistrationMethodNamePatterns`

> Regex replacement patterns (format `"pattern=>replacement"`) applied to handler or saga type names to customize the generated registration method names. Patterns are applied in order and the first match wins.

Reviewer: I think here we have to describe how handler class name is turned into the registration method, and the instances in which that might turn out silly and annoying, and then show how the patterns can fix those, otherwise it all just sounds very abstract and hand-wavy.

Author: Aren't the samples below given enough coverage of that aspects?

Author: Gave it a fix

Examples:

- `Handler$=>Register` transforms `OrderShippedHandler` to `OrderShippedRegister`.
- `^(.*)Handler$=>Add$1` transforms `OrderShippedHandler` to `AddOrderShipped`.
- `^(.*)Saga$=>Register$1Saga` transforms `OrderShippingSaga` to `RegisterOrderShippingSaga`.

#### Visibility control

The source generator automatically matches the visibility of the generated extension methods to the visibility of the attributed `partial class`. Declaring the class as `internal` hides the registration methods from the public API.

#### Default behavior

> When the attribute is not applied, the source generator uses the assembly name as the entry point and a default naming convention.

Reviewer: Could be confusing to list the default last. Say what happens by default and _then_ describe how to advancedly configure it.

#### Relationship to other attributes

`[HandlerRegistryExtensions]` works alongside `[Handler]` (for marking message handlers) and `[Saga]` (for marking sagas) to enable source generation.

## Manual registration

When source generation is not available or when only a few components need to be registered, individual types can be added explicitly.

### Register a message handler

Use `AddHandler<THandler>()` to register a message handler:

snippet: RegisterHandlerManually

### Register a saga

Use `AddSaga<TSaga>()` to register a saga:

snippet: RegisterSagaManually

### Enable a feature

Use `EnableFeature<TFeature>()` to enable a feature:

snippet: EnableFeatureManually

### Register an installer

> Use `AddInstaller<TInstaller>()` to register an installer:

Reviewer 2: I assume we also documents this in the specific feature and installer docs

Author: Yes there is a slight redundancy now. We could also remove those sections here
